### PR TITLE
Gurobi write model to file.

### PR DESCRIPTION
**Pull request by hongkai-dai**
Also can compute IIS when the model is infeasible.

@TobiaMarcucci this feature can be useful when you debug infeasible Gurobi problem.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@avalenzu for feature review please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-Mf5Qp_M4sw4rreCIjzH:bll7v0f)* status: LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu)

**Review by avalenzu — COMMENTED** [review]
Review pass done. A couple of naming requests.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-Mf8FXCn5WigYn-YjxeJ:b7i0q3e)* status: 2 unresolved discussions, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai)

___
*[solvers/gurobi_solver.h, line 66 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15438#-Mf8EtPy-bswCXi5pDyt:-Mf8EtPy-bswCXi5pDyu:b-4hufqv) ([raw file](https://github.com/robotlocomotion/drake/blob/03a2e1256eb0d8dd6750858d6ac3bfaea14828d5/solvers/gurobi_solver.h#L66)):*
> ```objc
>    * including the file extension. file_name="" means don't write to file.
>    */
>   void write_to_file(const std::string& file_name) { write_file_ = file_name; }
> ```

This name is somewhat misleading, as this method doesn't write to a file. Consider using `set_write_file` or `set_model_output_file` (with a corresponding change to the member variable name).
___
*[solvers/gurobi_solver.h, line 78 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15438#-Mf8ESzD0LvaW_VKkj-n:-Mf8ESzE2kJCnsD9xHVB:bll4t1y) ([raw file](https://github.com/robotlocomotion/drake/blob/03a2e1256eb0d8dd6750858d6ac3bfaea14828d5/solvers/gurobi_solver.h#L78)):*
> ```objc
>    * @param iis_flag Set to true so that when the problem is infeasible
>    */
>   void compute_iis(bool iis_flag) { iis_flag_ = iis_flag; }
> ```

This name is somewhat misleading, as this method doesn't do any computation. Consider using `set_iss_flag`.

**Comment by jwnimmer-tri** [comment]
For solver-specific configuration that ends up calling into the solver back-end's direct API with a specific function name (vs the back-end's generic named-options API), we could imagine still using Drake's SolverOptions for that, with a Drake-specific chosen name for the setting.

For example, maybe:
```
prog.SetSolverOption(GurobiSolver::id(), "drake:GRBcomputeIIS", 1);
prog.SetSolverOption(GurobiSolver::id(), "drake:GRBwrite", "model.mps");
```

That might require less API boilerplate (and new pydrake bindings), and make it easier to pass the option down into the solver, in case the user don't have pointer access to the chosen solver instance.

**Review by hongkai-dai — COMMENTED** [review]
I rewrite this function to use the SolverOption approach that Jeremy suggested.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-Mf9YHnm0VLDX8seyQ8N:beg8ay7)* status: 2 unresolved discussions, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu)

___
*[solvers/gurobi_solver.h, line 66 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15438#-Mf8EtPy-bswCXi5pDyt:-Mf9YNVO-kV-vYQdVtOf:b-z9pbo0) ([raw file](https://github.com/robotlocomotion/drake/blob/03a2e1256eb0d8dd6750858d6ac3bfaea14828d5/solvers/gurobi_solver.h#L66)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

This name is somewhat misleading, as this method doesn't write to a file. Consider using `set_write_file` or `set_model_output_file` (with a corresponding change to the member variable name).
</blockquote></details>

Done. I rewrite the function to use the SolverOption approach that Jeremy suggested.
___
*[solvers/gurobi_solver.h, line 78 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15438#-Mf8ESzD0LvaW_VKkj-n:-Mf9YSFO25jQ61fi1AhP:bg60x3x) ([raw file](https://github.com/robotlocomotion/drake/blob/03a2e1256eb0d8dd6750858d6ac3bfaea14828d5/solvers/gurobi_solver.h#L78)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

This name is somewhat misleading, as this method doesn't do any computation. Consider using `set_iss_flag`.
</blockquote></details>

Done. Ditto.

**Review by avalenzu — COMMENTED** [review]
Oooh, I like this! :lgtm: 

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-Mf9puPe9io9bQWZuO5e:bnplbef)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai)

**Review by hongkai-dai — COMMENTED** [review]
+@jwnimmer-tri for platform review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-MfE7r9p154gqTZwRv7G:buxscyd)* status: LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @hongkai-dai)

___
*[solvers/gurobi_solver.cc, line 1063 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfE9eY35UAr47L4XaHo:-MfE9eY35UAr47L4XaHp:b-ywp33a) ([raw file](https://github.com/robotlocomotion/drake/blob/9b7893e56ae79c337453db164e0a61eb9494dca5/solvers/gurobi_solver.cc#L1063)):*
> ```cpp
> 
>   if (!error) {
>     const auto& int_options = merged_options.GetOptionsInt(id());
> ```

The options-handling code is too spread out to be safely maintainable in such a huge function.

Furthermore, if the options are provided incorrectly by the user, then we should produce that error while preparing the optimization, not later after we've already solved it.

I think the best fix for this would be to have function-local variables, like so:

```c++
bool compute_iis = false;
std::optional<std::string> grb_write;
```

They could be declared around L1004 above, and then the pre-existing loop `for (const auto& it : merged_options.GetOptionsInt(id()))` that is handling the options, that already has an "if" check for these magic strings, would be responsible for validating the options and changing the local variables to reflect the input.

Then here, after the optimization, we can just use the local variables to condition whether or not we call these extra Gurobi methods.
___
*[solvers/gurobi_solver.cc, line 1076 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEBZ_8EU4uwLVsRXJ5:-MfEBZ_8EU4uwLVsRXJ6:b-sc3ac9) ([raw file](https://github.com/robotlocomotion/drake/blob/9b7893e56ae79c337453db164e0a61eb9494dca5/solvers/gurobi_solver.cc#L1076)):*
> ```cpp
>         }
>       } else {
>         drake::log()->warn(
> ```

From recent PRs, my understanding was that we were going to communicate malformed options using exceptions, not logs, from now on?

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @avalenzu and @jwnimmer-tri)

___
*[solvers/gurobi_solver.cc, line 1063 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfE9eY35UAr47L4XaHo:-MfEMmXIEKi2jHYn__Zb:b-yh16l6) ([raw file](https://github.com/robotlocomotion/drake/blob/9b7893e56ae79c337453db164e0a61eb9494dca5/solvers/gurobi_solver.cc#L1063)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The options-handling code is too spread out to be safely maintainable in such a huge function.

Furthermore, if the options are provided incorrectly by the user, then we should produce that error while preparing the optimization, not later after we've already solved it.

I think the best fix for this would be to have function-local variables, like so:

```c++
bool compute_iis = false;
std::optional<std::string> grb_write;
```

They could be declared around L1004 above, and then the pre-existing loop `for (const auto& it : merged_options.GetOptionsInt(id()))` that is handling the options, that already has an "if" check for these magic strings, would be responsible for validating the options and changing the local variables to reflect the input.

Then here, after the optimization, we can just use the local variables to condition whether or not we call these extra Gurobi methods.
</blockquote></details>

Done. Good call.
___
*[solvers/gurobi_solver.cc, line 1076 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEBZ_8EU4uwLVsRXJ5:-MfEMnoi5Sh7G1d7Lqu_:bqq0vg5) ([raw file](https://github.com/robotlocomotion/drake/blob/9b7893e56ae79c337453db164e0a61eb9494dca5/solvers/gurobi_solver.cc#L1076)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

From recent PRs, my understanding was that we were going to communicate malformed options using exceptions, not logs, from now on?
</blockquote></details>

But here I just want to warn the user, instead of throwing an error (because Gurobi can still proceed with this wrong option, while setting the other options to the wrong value, for example setting DualReduction to a wrong value, would cause Gurobi to throw an error). Should I also use `throw` for warning?

**Review by avalenzu — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: 6 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[solvers/gurobi_solver.cc, line 1012 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEOgUU3ISsblEVfMSQ:-MfEOgUU3ISsblEVfMSR:bembx5z) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1012)):*
> ```cpp
>   for (const auto& it : merged_options.GetOptionsInt(id())) {
>     if (!error) {
>       if (it.first != "GRBcomputeIIS") {
> ```

nit The if-else choice here is a bit odd.  I suppose it's fine to leave it this way (it's not logically incorrect), but as soon as we have a second magic option name, it will have to be rewritten.

Here's what seems like a better pattern for the loop body, to me:
```c++
if (error) { break; }
if (it.first == "GRBcomputeIIS") {
  compute_iis = ... etc.
  ... etc.
  continue;
}
error = GRBsetintparam(model_env, it.first.c_str(), it.second);
```

Ditto below.
___
*[solvers/gurobi_solver.cc, line 1016 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEO7ER4cGy7H1ivnar:-MfEO7ES3IyyS9Pco6Al:b-700sp3) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1016)):*
> ```cpp
>       } else {
>         compute_iis = static_cast<bool>(it.second);
>         if (!(it.second== 0 || it.second== 1)) {
> ```

nit Missing whitespace between `it.second` and `==`.
___
*[solvers/gurobi_solver.cc, line 1017 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEPNeR8qRR_BMNre6I:-MfEPNeS3YvMvpHxhSvI:b-f76268) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1017)):*
> > From recent PRs, my understanding was that we were going to communicate malformed options using exceptions, not logs, from now on?
>
> But here I just want to warn the user, instead of throwing an error (because Gurobi can still proceed with this wrong option, while setting the other options to the wrong value, for example setting DualReduction to a wrong value, would cause Gurobi to throw an error). Should I also use throw for warning?

I think we need a coherent story to explain to a user what situations turn into `MathematicalProgramResult.is_success() == false`, vs what situations spam the console, vs what situations throw an exception.  I haven't yet heard or been able to think of any proposal for that question which is consistent with this implementation.

For predicates that are only under loose control of the user (e.g., asking the user to predict an upper bound on the iteration limit), an exception may be too harsh.

But in cases like this spot, where the user has obviously made a very direct and unwanted typo in their source code by setting a boolean option to a non-binary value, asking them to repair their source code before we proceed seems reasonable to me?

Or as another alternative, you could allow any value here, and use "zero vs non-zero" as the predicate for whether to turn on iis.  So "-1" and "2" would act the same as "1", without any remark.

In short, noting non-binary values is either important to us (in which case we should throw), or not important at all (in which case we should remain silent).

I would probably err on the side of however Gurobi itself treats boolean options.  If it errors when given non-0,1, they we should here also.  If it silently treats non-zero as true, we should here also.
___
*[solvers/gurobi_solver.cc, line 1031 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEOQbL9JZ5nqmn0pMP:-MfEOQbL9JZ5nqmn0pMQ:b-peii1j) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1031)):*
> ```cpp
>         error = GRBsetstrparam(model_env, it.first.c_str(), it.second.c_str());
>       } else {
>         grb_write = it.second;
> ```

nit Consider skipping this assignment when `it.second` is empty.  That would simplify the logic below (at least, that's how I perceive it).
___
*[solvers/gurobi_solver.cc, line 1075 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEPkNh4oZvQ5lYoCh5:-MfEPkNh4oZvQ5lYoCh6:bcx0sw9) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1075)):*
> ```cpp
> 
>   if (!error) {
>     if (compute_iis == 1) {
> ```

nit The `compute_iis` is a bool, we shouldn't use `== 1`.
___
*[solvers/gurobi_solver.cc, line 1088 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEQ0WS9hYivuLSLqzb:-MfEQ0WS9hYivuLSLqzc:bmee4f8) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1088)):*
> ```cpp
>       error = GRBwrite(model, grb_write.value().c_str());
>       if (error) {
>       const std::string gurobi_version =
> ```

nit Wrong indentation.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri)

___
*[solvers/gurobi_solver.cc, line 1012 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEOgUU3ISsblEVfMSQ:-MfG9QR_9Nzjg-bmM-Io:b-h5g1tk) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1012)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit The if-else choice here is a bit odd.  I suppose it's fine to leave it this way (it's not logically incorrect), but as soon as we have a second magic option name, it will have to be rewritten.

Here's what seems like a better pattern for the loop body, to me:
```c++
if (error) { break; }
if (it.first == "GRBcomputeIIS") {
  compute_iis = ... etc.
  ... etc.
  continue;
}
error = GRBsetintparam(model_env, it.first.c_str(), it.second);
```

Ditto below.
</blockquote></details>

Done. Good call
___
*[solvers/gurobi_solver.cc, line 1017 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfEPNeR8qRR_BMNre6I:-MfG9wKu2VUmhYWioitV:bp1mu5z) ([raw file](https://github.com/robotlocomotion/drake/blob/3a5db31ea5249b15ec5f0fc77bdb1cb59aa88778/solvers/gurobi_solver.cc#L1017)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> > From recent PRs, my understanding was that we were going to communicate malformed options using exceptions, not logs, from now on?
>
> But here I just want to warn the user, instead of throwing an error (because Gurobi can still proceed with this wrong option, while setting the other options to the wrong value, for example setting DualReduction to a wrong value, would cause Gurobi to throw an error). Should I also use throw for warning?

I think we need a coherent story to explain to a user what situations turn into `MathematicalProgramResult.is_success() == false`, vs what situations spam the console, vs what situations throw an exception.  I haven't yet heard or been able to think of any proposal for that question which is consistent with this implementation.

For predicates that are only under loose control of the user (e.g., asking the user to predict an upper bound on the iteration limit), an exception may be too harsh.

But in cases like this spot, where the user has obviously made a very direct and unwanted typo in their source code by setting a boolean option to a non-binary value, asking them to repair their source code before we proceed seems reasonable to me?

Or as another alternative, you could allow any value here, and use "zero vs non-zero" as the predicate for whether to turn on iis.  So "-1" and "2" would act the same as "1", without any remark.

In short, noting non-binary values is either important to us (in which case we should throw), or not important at all (in which case we should remain silent).

I would probably err on the side of however Gurobi itself treats boolean options.  If it errors when given non-0,1, they we should here also.  If it silently treats non-zero as true, we should here also.
</blockquote></details>

When the integer value should be either 0 or 1, and the user provide a non-binary value, Gurobi will return GRB_VALUE_OUT_OF_RANGE error. https://www.gurobi.com/documentation/9.1/refman/dualreductions.html.

I changed to throw a runtime error here.

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438#-:-MfGmWM30Dyqe7hHsqht:bnfp4nl)* status: 2 unresolved discussions (waiting on @hongkai-dai)

___
*[solvers/gurobi_solver.cc, line 1024 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfGmLgqBY7rn2aCfB5U:-MfGmLgqBY7rn2aCfB5V:bxkg49g) ([raw file](https://github.com/robotlocomotion/drake/blob/f59a52e37fdecad8b0fae1d0fe441d27e0d55a72/solvers/gurobi_solver.cc#L1024)):*
> ```cpp
>         error = GRBsetintparam(model_env, it.first.c_str(), it.second);
>         if (error) {
>           break;
> ```

nit This "break" doesn't accomplish much -- we already have an `if (!error)` guarding this whole loop body each time?
___
*[solvers/gurobi_solver.cc, line 1091 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15438#-MfGmIYF2pDcY_ugY_sR:-MfGmIYF2pDcY_ugY_sS:bbc460g) ([raw file](https://github.com/robotlocomotion/drake/blob/f59a52e37fdecad8b0fae1d0fe441d27e0d55a72/solvers/gurobi_solver.cc#L1091)):*
> ```cpp
>   }
>   if (!error) {
>     if (grb_write.has_value() && grb_write.value() != "") {
> ```

nit We don't need the `!= ""` here anymore, I think?

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15438)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),avalenzu (waiting on @hongkai-dai)
